include/pybind11/cast.h

```diff
@@ -915,10 +937,15 @@ template <typename type> class type_caster_base : public type_caster_generic {
             nullptr, nullptr, holder);
     }
 
-    template <typename T> using cast_op_type = detail::cast_op_type<T>;
+    // allow moving of custom types
+    template <typename T> using cast_op_type = detail::movable_cast_op_type<T>;
 
     operator itype*() { return (type *) value; }
     operator itype&() { if (!value) throw reference_cast_error(); return *((itype *) value); }
+    // move caster, actually triggering a move construction from value
+    operator itype() { if (!value) throw reference_cast_error(); return std::move(*((itype *) value)); }
+    // fallback move caster, just passing rvalue reference as is
+    operator itype&&() { if (!value) throw reference_cast_error(); return std::move(*((itype *) value)); }
```

Comment:
This overload set seems weird. With a maybe too high WTF per line of code ratio. I'm not saying it's wrong, but I'm sure everyone would benefit from some clarifications. For context:


> - `f(T&&)` triggers move semantics for `T` (if available)
> - `f(T)` triggers copy semantics for `T` (and falls back to move semantics if copying is not supported)
> - `f(T&)` uses lvalue-references as before.


Here are the questions:

1. What's the difference between `operator itype` and `operator itype&&`? Both look the same.
2. If we want `f(T)` to be a copy (which seems reasonable), why is `operator itype` using `std::move()`?
3. Should the operators be ref-qualified, as in `operator itype&&() &&`?

Reply:
> 1. What's the difference between `operator itype` and `operator itype&&`? Both look the same.

Both variants are needed to actually trigger a move construction in both of the following cases:
- for a function `foo(T&&)`, explicitly trigger a move construction via `operator itype`
  Just passing the rvalue reference, wouldn't trigger a move construction!
- for a function `foo(T)` with a non-copy-constructible type, implicitly trigger a move construction
  via `operator itype&&`

Introducing the extra cast operator is just a technical trick to actually achieve the correct move-construction behavior.
I also tried [including this logic into `cast_op<T>()`](https://github.com/rhaschke/pybind11/blob/36ba36e892aeac3adb48e59d5edc313be2521713/include/pybind11/cast.h#L976-L981) instead. While this approach will remove the need to define the redundant `operator itype()`, my naive implementation returns a reference to a local temporary. Alternatively changing the return type based on the two cases, will make `caster::cast_op_type` inconsistent to what `cast_op` actually returns. Maybe, you, @bstaletic, have a better idea to implement it there.

> 2. If we want `f(T)` to be a copy (which seems reasonable), why is `operator itype` using `std::move()`?

In C++, the `f(T)` uses a copy constructor for lvalue references and a move constructor for rvalue references.
Returning `std::move()` from `operator itype()` essentially yields `itype(std::move(arg))`, which realizes the move construction.

> 3. Should the operators be ref-qualified, as in `operator itype&&() &&`?

I think this extra rvalue-qualifier was used in the past to restrict the move operation to situations where the caster itself is an rvalue. However, due to [`cast_op<T>(caster&&)`](https://github.com/pybind/pybind11/pull/2047/files#diff-b8e97eea5f84a84248db1bc72afaccfd8ba28cad023917d52bdc6bbc3669cd74L953-L957), the cast_op_type is always an rvalue reference if the caster is an rvalue reference. There is no need to require an rvalue caster also for the actual casting operation.

I improved the source code comments, see f93d0f01f25472dd40f542d450f87fe518ae952f.